planner/core/plan.go

```diff
@@ -126,11 +131,61 @@ func optimizeByShuffle4Window(pp *PhysicalWindow, ctx sessionctx.Context) *Physi
 	}
 	reqProp := &property.PhysicalProperty{ExpectedCnt: math.MaxFloat64}
 	shuffle := PhysicalShuffle{
-		Concurrency:  concurrency,
-		Tail:         tail,
-		DataSource:   dataSource,
-		SplitterType: PartitionHashSplitterType,
-		HashByItems:  byItems,
+		Concurrency:      concurrency,
+		Tails:            []PhysicalPlan{tail},
+		DataSources:      []PhysicalPlan{dataSource},
+		SplitterType:     PartitionHashSplitterType,
+		HashByItemArrays: [][]expression.Expression{byItems},
+	}.Init(ctx, pp.statsInfo(), pp.SelectBlockOffset(), reqProp)
+	return shuffle
+}
+
+func optimizeByShuffle4MergeJoin(pp *PhysicalMergeJoin, ctx sessionctx.Context) *PhysicalShuffle {
+	concurrency := 4
+	// concurrency := ctx.GetSessionVars().WindowConcurrency()
+	// if concurrency <= 1 {
+	// 	return nil
+	// }
+
+	children := pp.Children()
+	dataSources := make([]PhysicalPlan, len(children))
+	tails := make([]PhysicalPlan, len(children))
+
+	for i := range children {
+		sort, ok := children[i].(*PhysicalSort)
+		if !ok {
+			// Multi-thread executing on SORTED data source is not effective enough by current implementation.
+			// TODO: Implement a better one.
+			return nil
+		}
+		tails[i], dataSources[i] = sort, sort.Children()[0]
+	}
+
+	// partitionBy := make([]*expression.Column, 0, len(pp.PartitionBy))
+	// for _, item := range pp.PartitionBy {
+	// 	partitionBy = append(partitionBy, item.Col)
+	// }
+	// NDV := int(getCardinality(partitionBy, dataSource.Schema(), dataSource.statsInfo()))
+	// if NDV <= 1 {
+	// 	return nil
+	// }
+	// concurrency = mathutil.Min(concurrency, NDV)
+
+	leftByItemArray := make([]expression.Expression, 0, len(pp.LeftJoinKeys))
+	for _, col := range pp.LeftJoinKeys {
+		leftByItemArray = append(leftByItemArray, col)
+	}
+	rightByItemArray := make([]expression.Expression, 0, len(pp.RightJoinKeys))
+	for _, col := range pp.RightJoinKeys {
+		rightByItemArray = append(rightByItemArray, col)
```

Comment:
It's better to clone them `leftByItemArray = append(leftByItemArray, col.Clone())`